core/shared/src/main/scala/zio/internal/FiberRuntime.scala

```diff
@@ -1094,30 +1097,29 @@ final class FiberRuntime[E, A](fiberId: FiberId.Runtime, fiberRefs0: FiberRefs,
             case fold: FoldZIO[Any, Any, Any, Any, Any] =>
               updateLastTrace(fold.trace)
 
-              stackIndex = pushStackFrame(fold, stackIndex)
+              @inline def exitFailure(cause: Cause[Any]): ZIO.Erased =
+                if (shouldInterrupt()) Exit.Failure(cause.stripFailures) else fold.failureK(cause)
 
-              val result = runLoop(fold.first, stackIndex, stackIndex, currentDepth + 1, ops)
-              ops += 1
+              val first = fold.first
 
-              if (null eq result)
-                return null
+              if (first eq ZIO.unit) cur = fold.successK(())
+              else if (first.isInstanceOf[Success[Any]]) cur = fold.successK(first.asInstanceOf[Success[Any]].value)
+              else if (first.isInstanceOf[Failure[Any]]) cur = exitFailure(first.asInstanceOf[Failure[Any]].cause)
```

Comment:
I think let's keep `Fold` as is to make it easier on the C2 compiler to optimize the branches. I run the ZIO test suite using `if (fold.first eq ZIO.unit) println("fold.first is ZIO.unit")` and it was printed 50 times only, so perhaps the shortcut here might end up making things worse overall

Reply:
> I run the ZIO test suite using if (fold.first eq ZIO.unit) println("fold.first is ZIO.unit") and it was printed 50 times only, so perhaps the shortcut here might end up making things worse overall

Don't you think this might happen more often in a real codebase? 🤔
If it's happening 50 times just in the unit tests, I'd expect it more in an actual code, no?

Reply:
> If it's happening 50 times just in the unit tests, I'd expect it more in an actual code, no?

I find it a bit hard to believe. There are 2,500 tests (plus `nonFlaky` annotations) plus the test executor code etc. A single test can contain hundreds/thousands of invocations of the runloop. I find it a bit hard to believe that user code will somehow result in having lots of cases where the this branch would be invoked

Reply:
Fair enough. I made the change.

Reply:
It might be interesting to explore (in another PR) optimisation the handling of this kind of case:
<img width="536" alt="image" src="https://github.com/user-attachments/assets/b382c988-c056-4663-9907-ceb0d29c4573" />

This is, I think, very common in the ZIO code to use `.foldZIO(success = ZIO.successFn, ...)`

Don't you think? 🤔